lib/spack/spack/build_environment.py

```diff
@@ -322,13 +322,32 @@ def set_build_environment_variables(pkg, env, dirty=False):
         extra_rpaths = ':'.join(compiler.extra_rpaths)
         env.set('SPACK_COMPILER_EXTRA_RPATHS', extra_rpaths)
 
-    # Add bin directories from dependencies to the PATH for the build.
-    bin_dirs = reversed(
-        [d.prefix.bin for d in pkg.spec.dependencies(deptype='build')
-         if os.path.isdir(d.prefix.bin)])
+    # Function to add a directory to a list only if
+    # that directory exists and is not already in the list
+    def add_dir_to_list(bin_dir, list):
+        if os.path.isdir(bin_dir):
+            if bin_dir not in list:
+                list.append(bin_dir)
+
+    # Create a list which will contain the bin directories
+    bin_dirs = []
+
+    # Add all bin directories which build dependencies and their
+    # run dependencies require.
+    for bdep in pkg.spec.dependencies(deptype='build'):
+        add_dir_to_list("%s/bin" % bdep.prefix, bin_dirs)
+        for brdep in bdep.traverse(root=False, deptype='run'):
+            add_dir_to_list("%s/bin" % brdep.prefix, bin_dirs)
+
+    # Add all bin directories which run dependencies require.
+    for rdep in pkg.spec.traverse(root=False, deptype='run'):
+        add_dir_to_list("%s/bin" % rdep.prefix, bin_dirs)
```

Comment:
IMO if the user wants a package to be part of the path when doing an install they should mark that package as a build dependency. This may be useful later since Spack in the future will try to concretize build dependencies separately (so if a package is 'run' but not 'build' that can help it avoid conflicts).